cpp/log/cert.cc

```diff
@@ -546,12 +588,311 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
+
+        if (IsRedactedHost(dns_name)) {
+          if (!IsValidRedactedHost(dns_name)) {
+            LOG(WARNING) << "Invalid redacted host: " << dns_name;
+            sk_GENERAL_NAME_free(subject_alt_names);
+            return Cert::FALSE;
+          }
+
+          redacted_name_count++;
+        }
+      }
+    }
+
+    sk_GENERAL_NAME_free(subject_alt_names);
+  }
+
+  // The next stage of validation is that if the subject name CN exists it
+  // must match the first DNS id and have the same labels redacted
+  // TODO: Confirm it's valid to not have a CN.
+  X509_NAME* name = X509_get_subject_name(x509_);
+
+  if (name == nullptr) {
+    LOG(ERROR) << "Missing X509 subject name";
+    return Cert::ERROR;
+  }
+
+  int name_pos = -1;
+
+  name_pos = X509_NAME_get_index_by_NID(name, NID_commonName, name_pos);
+
+  if (name_pos >= 0) {
+    X509_NAME_ENTRY* name_entry = X509_NAME_get_entry(name, name_pos);
+
+    if (name_entry != NULL) {
+      ASN1_STRING *subject_name_asn1 = X509_NAME_ENTRY_get_data(name_entry);
+
+      if (subject_name_asn1 == nullptr) {
+        LOG(WARNING) << "Missing subject name";
+        // TODO: Check this is correct behaviour. Is it OK to not have
+        // a subject?
+      } else {
+        Cert::Status cn_status;
+        common_name = ASN1ToStringAndCheckForNulls(subject_name_asn1,
+                                                   "CN", cn_status);
+
+        if (cn_status != TRUE) {
+          return cn_status;
+        }
+      }
+    }
+  }
+
+  // If both a subject CN and DNS ids are present in the cert then the
+  // first DNS id must exactly match the CN
+  if (!dns_alt_names.empty() && common_name.length() > 0) {
+    if (dns_alt_names[0] != common_name) {
+      LOG(WARNING) << "CN " << common_name << " does not match DNS.0 "
+          << dns_alt_names[0];
+      return Cert::FALSE;
+    }
+  }
+
+  // The attempted redaction passes host validation. Stage two is checking
+  // that the required extensions are present and specified correctly if
+  // we found any redacted names. First though if nothing is redacted
+  // then the rest of the rules need not be applied
+  if (redacted_name_count == 0 && !IsRedactedHost(common_name)) {
+    return Cert::TRUE;
+  }
+
+  // If we reach here then the RFC says the CT redaction count extension
+  // MUST BE present.
+  X509_EXTENSION* exty;
+  if (GetExtension(NID_ctPrecertificateRedactedLabelCount,
+                   &exty) != Cert::TRUE) {
+    LOG(WARNING) << "Required CT redaction count extension missing from cert";
+    return Cert::FALSE;
+  }
+
+  // Unpack the extension contents, which should be SEQUENCE OF INTEGER
+  STACK_OF(ASN1_INTEGER)* integers = (STACK_OF(ASN1_INTEGER)*)
+      ASN1_seq_unpack_ASN1_INTEGER(exty->value->data, exty->value->length,
+                                   d2i_ASN1_INTEGER, ASN1_INTEGER_free);
+
+  if (integers != NULL) {
+    const int num_integers = sk_ASN1_INTEGER_num(integers);
```

Comment:
It somehow looks wrong to me, like it's calling a method. Similarly in java you can't write if (ptr), which is why I do it.